File: sdk/core/unity/Runtime/Core/AdaptiveConcurrency.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using UnityEngine;

namespace RivalApexMediation.Core.Threading
{
    /// <summary>
    /// AdaptiveConcurrency - Adaptive task scheduler sizing based on device capabilities.
    /// 
    /// This class provides adaptive concurrency management for Unity that:
    /// - Detects CPU core count and platform capabilities
    /// - Sizes task limits appropriately for the device
    /// - Provides separate scheduling for CPU-bound and I/O-bound operations
    /// - Integrates with Unity's main thread for UI updates
    /// </summary>
    public sealed class AdaptiveConcurrency : IDisposable
    {
        private static AdaptiveConcurrency _instance;
        private static readonly object _lock = new object();

        /// <summary>
        /// Gets the singleton instance of AdaptiveConcurrency.
        /// </summary>
        public static AdaptiveConcurrency Instance
        {
            get
            {
                if (_instance == null)
                {
                    lock (_lock)
                    {
                        if (_instance == null)
                        {
                            _instance = new AdaptiveConcurrency();
                        }
                    }
                }
                return _instance;
            }
        }

        // Configuration constants
        private const int MinConcurrentTasks = 2;
        private const int MaxBackgroundTasks = 8;
        private const int MaxNetworkTasks = 16;
        private const int MaxComputeTasks = 4;

        /// <summary>
        /// Detected CPU information for the current device.
        /// </summary>
        public struct CpuInfo
        {
            public int TotalCores;
            public string Platform;
            public int SystemMemoryMB;
            public bool IsLowEndDevice;
            public GraphicsDeviceType GraphicsType;
        }

        /// <summary>
        /// Pool configuration for different workload types.
        /// </summary>
        public struct PoolConfig
        {
            public int MaxConcurrency;
            public ThreadPriority Priority;
        }

        /// <summary>
        /// Graphics device type enumeration.
        /// </summary>
        public enum GraphicsDeviceType
        {
            Unknown,
            OpenGL,
            Metal,
            Vulkan,
            DirectX
        }

        // Properties
        public CpuInfo DeviceInfo { get; private set; }

        // Semaphores for limiting concurrent operations
        private readonly SemaphoreSlim _backgroundSemaphore;
        private readonly SemaphoreSlim _networkSemaphore;
        private readonly SemaphoreSlim _computeSemaphore;

        // Metrics
        private int _taskCount;
        private int _completedTaskCount;
        private int _rejectedTaskCount;
        
        // Main thread synchronization context
        private readonly SynchronizationContext _mainThreadContext;
        
        private bool _disposed;

        private AdaptiveConcurrency()
        {
            DeviceInfo = DetectCpuInfo();
            
            // Create semaphores based on device capabilities
            var backgroundConfig = GetBackgroundPoolConfig();
            var networkConfig = GetNetworkPoolConfig();
            var computeConfig = GetComputePoolConfig();
            
            _backgroundSemaphore = new SemaphoreSlim(backgroundConfig.MaxConcurrency, backgroundConfig.MaxConcurrency);
            _networkSemaphore = new SemaphoreSlim(networkConfig.MaxConcurrency, networkConfig.MaxConcurrency);
            _computeSemaphore = new SemaphoreSlim(computeConfig.MaxConcurrency, computeConfig.MaxConcurrency);
            
            // Capture main thread context for callbacks
            _mainThreadContext = SynchronizationContext.Current;
            
            Debug.Log($"[AdaptiveConcurrency] Initialized: {DeviceInfo.TotalCores} cores, Platform: {DeviceInfo.Platform}, LowEnd: {DeviceInfo.IsLowEndDevice}");
        }

        /// <summary>
        /// Gets the recommended pool configuration for background work.
        /// </summary>
        public PoolConfig GetBackgroundPoolConfig()
        {
            int maxConcurrency = DeviceInfo.IsLowEndDevice 
                ? Math.Min(2, DeviceInfo.TotalCores)
                : Math.Min(DeviceInfo.TotalCores - 1, MaxBackgroundTasks);
            
            return new PoolConfig
            {
                MaxConcurrency = Math.Max(MinConcurrentTasks, maxConcurrency),
                Priority = ThreadPriority.BelowNormal
            };
        }

        /// <summary>
        /// Gets the recommended pool configuration for network I/O.
        /// </summary>
        public PoolConfig GetNetworkPoolConfig()
        {
            int maxConcurrency = DeviceInfo.IsLowEndDevice
                ? Math.Min(4, DeviceInfo.TotalCores * 2)
                : Math.Min(DeviceInfo.TotalCores * 4, MaxNetworkTasks);
            
            return new PoolConfig
            {
                MaxConcurrency = Math.Max(MinConcurrentTasks, maxConcurrency),
                Priority = ThreadPriority.Normal
            };
        }

        /// <summary>
        /// Gets the recommended pool configuration for compute-intensive work.
        /// </summary>
        public PoolConfig GetComputePoolConfig()
        {
            int maxConcurrency = DeviceInfo.IsLowEndDevice
                ? 1
                : Math.Min(DeviceInfo.TotalCores / 2, MaxComputeTasks);
            
            return new PoolConfig
            {
                MaxConcurrency = Math.Max(1, maxConcurrency),
                Priority = ThreadPriority.AboveNormal
            };
        }

        /// <summary>
        /// Executes a background task with concurrency limiting.
        /// </summary>
        public async Task ExecuteBackgroundAsync(Func<Task> task, CancellationToken cancellationToken = default)
        {
            Interlocked.Increment(ref _taskCount);
            
            try
            {
                await _backgroundSemaphore.WaitAsync(cancellationToken);
                try
                {
                    await task();
                }
                finally
                {
                    _backgroundSemaphore.Release();
                    Interlocked.Increment(ref _completedTaskCount);
                }
            }
            catch (OperationCanceledException)
            {
                Interlocked.Increment(ref _rejectedTaskCount);
                throw;
            }
        }

        /// <summary>
        /// Executes a background task with concurrency limiting and returns a result.
        /// </summary>
        public async Task<T> ExecuteBackgroundAsync<T>(Func<Task<T>> task, CancellationToken cancellationToken = default)
        {
            Interlocked.Increment(ref _taskCount);
            
            try
            {
                await _backgroundSemaphore.WaitAsync(cancellationToken);
                try
                {
                    return await task();
                }
                finally
                {
                    _backgroundSemaphore.Release();
                    Interlocked.Increment(ref _completedTaskCount);
                }
            }
            catch (OperationCanceledException)
            {
                Interlocked.Increment(ref _rejectedTaskCount);
                throw;
            }
        }

        /// <summary>
        /// Executes a network I/O task with concurrency limiting.
        /// </summary>
        public async Task ExecuteNetworkAsync(Func<Task> task, CancellationToken cancellationToken = default)
        {
            Interlocked.Increment(ref _taskCount);
            
            try
            {
                await _networkSemaphore.WaitAsync(cancellationToken);
                try
                {
                    await task();
                }
                finally
                {
                    _networkSemaphore.Release();
                    Interlocked.Increment(ref _completedTaskCount);
                }
            }
            catch (OperationCanceledException)
            {
                Interlocked.Increment(ref _rejectedTaskCount);
                throw;
            }
        }

        /// <summary>
        /// Executes a network I/O task with concurrency limiting and returns a result.
        /// </summary>
        public async Task<T> ExecuteNetworkAsync<T>(Func<Task<T>> task, CancellationToken cancellationToken = default)
        {
            Interlocked.Increment(ref _taskCount);
            
            try
            {
                await _networkSemaphore.WaitAsync(cancellationToken);
                try
                {
                    return await task();
                }
                finally
                {
                    _networkSemaphore.Release();
                    Interlocked.Increment(ref _completedTaskCount);
                }
            }
            catch (OperationCanceledException)
            {
                Interlocked.Increment(ref _rejectedTaskCount);
                throw;
            }
        }

        /// <summary>
        /// Executes a compute-intensive task with concurrency limiting.
        /// </summary>
        public async Task ExecuteComputeAsync(Func<Task> task, CancellationToken cancellationToken = default)
        {
            Interlocked.Increment(ref _taskCount);
            
            try
            {
                await _computeSemaphore.WaitAsync(cancellationToken);
                try
                {
                    await task();
                }
                finally
                {
                    _computeSemaphore.Release();
                    Interlocked.Increment(ref _completedTaskCount);
                }
            }
            catch (OperationCanceledException)
            {
                Interlocked.Increment(ref _rejectedTaskCount);
                throw;
            }
        }

        /// <summary>
        /// Executes a compute-intensive task with concurrency limiting and returns a result.
        /// </summary>
        public async Task<T> ExecuteComputeAsync<T>(Func<Task<T>> task, CancellationToken cancellationToken = default)
        {
            Interlocked.Increment(ref _taskCount);
            
            try
            {
                await _computeSemaphore.WaitAsync(cancellationToken);
                try
                {
                    return await task();
                }
                finally
                {
                    _computeSemaphore.Release();
                    Interlocked.Increment(ref _completedTaskCount);
                }
            }
            catch (OperationCanceledException)
            {
                Interlocked.Increment(ref _rejectedTaskCount);
                throw;
            }
        }

        /// <summary>
        /// Executes a synchronous task on a background thread.
        /// </summary>
        public Task ExecuteOnThreadPool(Action action)
        {
            Interlocked.Increment(ref _taskCount);
            
            return Task.Run(() =>
            {
                try
                {
                    action();
                }
                finally
                {
                    Interlocked.Increment(ref _completedTaskCount);
                }
            });
        }

        /// <summary>
        /// Executes a callback on the main Unity thread.
        /// </summary>
        public void ExecuteOnMainThread(Action action)
        {
            if (_mainThreadContext != null)
            {
                _mainThreadContext.Post(_ => action(), null);
            }
            else
            {
                // Fallback if no synchronization context is available
                action();
            }
        }

        /// <summary>
        /// Gets current concurrency statistics.
        /// </summary>
        public Dictionary<string, object> GetStats()
        {
            return new Dictionary<string, object>
            {
                { "cpuCores", DeviceInfo.TotalCores },
                { "platform", DeviceInfo.Platform },
                { "isLowEndDevice", DeviceInfo.IsLowEndDevice },
                { "systemMemoryMB", DeviceInfo.SystemMemoryMB },
                { "backgroundSemaphore", new Dictionary<string, int>
                    {
                        { "available", _backgroundSemaphore.CurrentCount },
                        { "max", GetBackgroundPoolConfig().MaxConcurrency }
                    }
                },
                { "networkSemaphore", new Dictionary<string, int>
                    {
                        { "available", _networkSemaphore.CurrentCount },
                        { "max", GetNetworkPoolConfig().MaxConcurrency }
                    }
                },
                { "computeSemaphore", new Dictionary<string, int>
                    {
                        { "available", _computeSemaphore.CurrentCount },
                        { "max", GetComputePoolConfig().MaxConcurrency }
                    }
                },
                { "totalTasksSubmitted", _taskCount },
                { "totalTasksCompleted", _completedTaskCount },
                { "totalTasksRejected", _rejectedTaskCount }
            };
        }

        private static CpuInfo DetectCpuInfo()
        {
            int cores = SystemInfo.processorCount;
            int memoryMB = SystemInfo.systemMemorySize;
            string platform = Application.platform.ToString();
            
            // Determine if this is a low-end device
            bool isLowEnd = cores <= 2 || memoryMB < 2048;
            
            // Detect graphics type
            GraphicsDeviceType graphicsType = GraphicsDeviceType.Unknown;
            var unityGraphicsType = SystemInfo.graphicsDeviceType;
            
            if (unityGraphicsType.ToString().Contains("OpenGL"))
                graphicsType = GraphicsDeviceType.OpenGL;
            else if (unityGraphicsType.ToString().Contains("Metal"))
                graphicsType = GraphicsDeviceType.Metal;
            else if (unityGraphicsType.ToString().Contains("Vulkan"))
                graphicsType = GraphicsDeviceType.Vulkan;
            else if (unityGraphicsType.ToString().Contains("Direct"))
                graphicsType = GraphicsDeviceType.DirectX;
            
            return new CpuInfo
            {
                TotalCores = cores,
                Platform = platform,
                SystemMemoryMB = memoryMB,
                IsLowEndDevice = isLowEnd,
                GraphicsType = graphicsType
            };
        }

        public void Dispose()
        {
            if (_disposed) return;
            
            _backgroundSemaphore?.Dispose();
            _networkSemaphore?.Dispose();
            _computeSemaphore?.Dispose();
            
            _disposed = true;
        }

        /// <summary>
        /// Resets the singleton instance (for testing purposes).
        /// </summary>
        public static void Reset()
        {
            lock (_lock)
            {
                _instance?.Dispose();
                _instance = null;
            }
        }
    }
}

File: sdk/core/ios/Sources/Threading/AdaptiveConcurrency.swift
import Foundation
import os.log

/// AdaptiveConcurrency - Adaptive dispatch queue and operation queue sizing based on device capabilities.
///
/// This class provides adaptive concurrency management that:
/// - Detects CPU core count and processor type
/// - Sizes operation queues appropriately for the device
/// - Provides separate queues for CPU-bound and I/O-bound operations
/// - Uses QoS (Quality of Service) for priority management
public final class AdaptiveConcurrency {
    
    /// Shared singleton instance
    public static let shared = AdaptiveConcurrency()
    
    // MARK: - Configuration Constants
    
    private static let minConcurrentOperations = 2
    private static let maxBackgroundOperations = 8
    private static let maxNetworkOperations = 16
    private static let maxComputeOperations = 4
    
    // MARK: - CPU Information
    
    /// Detected CPU information for the current device.
    public struct CpuInfo {
        public let totalCores: Int
        public let performanceCores: Int
        public let efficiencyCores: Int
        public let isHeterogeneous: Bool
        public let processorType: String
    }
    
    /// Pool configuration for different workload types.
    public struct PoolConfig {
        public let maxConcurrentOperations: Int
        public let qualityOfService: QualityOfService
    }
    
    // MARK: - Properties
    
    /// Detected CPU information
    public let cpuInfo: CpuInfo
    
    /// Background operation queue for general background work
    public let backgroundQueue: OperationQueue
    
    /// Network operation queue for I/O-bound operations
    public let networkQueue: OperationQueue
    
    /// Compute operation queue for CPU-intensive work
    public let computeQueue: OperationQueue
    
    /// Dispatch queue for background work
    public let backgroundDispatchQueue: DispatchQueue
    
    /// Dispatch queue for network I/O
    public let networkDispatchQueue: DispatchQueue
    
    /// Dispatch queue for compute-intensive work
    public let computeDispatchQueue: DispatchQueue
    
    private let logger = Logger(subsystem: "com.rivalapexmediation.sdk", category: "AdaptiveConcurrency")
    
    // MARK: - Metrics
    
    private let metricsLock = NSLock()
    private var taskCount: Int = 0
    private var completedTaskCount: Int = 0
    
    // MARK: - Initialization
    
    private init() {
        self.cpuInfo = AdaptiveConcurrency.detectCpuInfo()
        
        // Create background queue
        let backgroundConfig = getBackgroundPoolConfig()
        self.backgroundQueue = OperationQueue()
        backgroundQueue.name = "com.rivalapexmediation.sdk.background"
        backgroundQueue.maxConcurrentOperationCount = backgroundConfig.maxConcurrentOperations
        backgroundQueue.qualityOfService = backgroundConfig.qualityOfService
        
        // Create network queue
        let networkConfig = getNetworkPoolConfig()
        self.networkQueue = OperationQueue()
        networkQueue.name = "com.rivalapexmediation.sdk.network"
        networkQueue.maxConcurrentOperationCount = networkConfig.maxConcurrentOperations
        networkQueue.qualityOfService = networkConfig.qualityOfService
        
        // Create compute queue
        let computeConfig = getComputePoolConfig()
        self.computeQueue = OperationQueue()
        computeQueue.name = "com.rivalapexmediation.sdk.compute"
        computeQueue.maxConcurrentOperationCount = computeConfig.maxConcurrentOperations
        computeQueue.qualityOfService = computeConfig.qualityOfService
        
        // Create dispatch queues with appropriate attributes
        self.backgroundDispatchQueue = DispatchQueue(
            label: "com.rivalapexmediation.sdk.background.dispatch",
            qos: .utility,
            attributes: .concurrent
        )
        
        self.networkDispatchQueue = DispatchQueue(
            label: "com.rivalapexmediation.sdk.network.dispatch",
            qos: .userInitiated,
            attributes: .concurrent
        )
        
        self.computeDispatchQueue = DispatchQueue(
            label: "com.rivalapexmediation.sdk.compute.dispatch",
            qos: .userInteractive,
            attributes: .concurrent
        )
        
        logger.info("AdaptiveConcurrency initialized: \(self.cpuInfo.totalCores) cores, heterogeneous: \(self.cpuInfo.isHeterogeneous)")
    }
    
    // MARK: - Pool Configuration
    
    /// Gets the recommended pool configuration for background work.
    public func getBackgroundPoolConfig() -> PoolConfig {
        let cores = cpuInfo.totalCores
        let maxOps = max(Self.minConcurrentOperations, min(cores - 1, Self.maxBackgroundOperations))
        return PoolConfig(
            maxConcurrentOperations: maxOps,
            qualityOfService: .utility
        )
    }
    
    /// Gets the recommended pool configuration for network I/O.
    public func getNetworkPoolConfig() -> PoolConfig {
        // Network I/O is mostly waiting, so we can have more concurrent operations
        let cores = cpuInfo.totalCores
        let maxOps = max(Self.minConcurrentOperations, min(cores * 4, Self.maxNetworkOperations))
        return PoolConfig(
            maxConcurrentOperations: maxOps,
            qualityOfService: .userInitiated
        )
    }
    
    /// Gets the recommended pool configuration for compute-intensive work.
    public func getComputePoolConfig() -> PoolConfig {
        // Compute work should use performance cores preferentially
        let cores = cpuInfo.isHeterogeneous ? cpuInfo.performanceCores : cpuInfo.totalCores
        let maxOps = max(Self.minConcurrentOperations, min(cores, Self.maxComputeOperations))
        return PoolConfig(
            maxConcurrentOperations: maxOps,
            qualityOfService: .userInteractive
        )
    }
    
    // MARK: - Task Execution
    
    /// Executes a background task using the background queue.
    public func executeBackground(_ work: @escaping () -> Void) {
        incrementTaskCount()
        backgroundQueue.addOperation {
            work()
            self.incrementCompletedCount()
        }
    }
    
    /// Executes a network I/O task using the network queue.
    public func executeNetwork(_ work: @escaping () -> Void) {
        incrementTaskCount()
        networkQueue.addOperation {
            work()
            self.incrementCompletedCount()
        }
    }
    
    /// Executes a compute-intensive task using the compute queue.
    public func executeCompute(_ work: @escaping () -> Void) {
        incrementTaskCount()
        computeQueue.addOperation {
            work()
            self.incrementCompletedCount()
        }
    }
    
    /// Executes a background task on the dispatch queue.
    public func dispatchBackground(_ work: @escaping () -> Void) {
        incrementTaskCount()
        backgroundDispatchQueue.async {
            work()
            self.incrementCompletedCount()
        }
    }
    
    /// Executes a network I/O task on the dispatch queue.
    public func dispatchNetwork(_ work: @escaping () -> Void) {
        incrementTaskCount()
        networkDispatchQueue.async {
            work()
            self.incrementCompletedCount()
        }
    }
    
    /// Executes a compute-intensive task on the dispatch queue.
    public func dispatchCompute(_ work: @escaping () -> Void) {
        incrementTaskCount()
        computeDispatchQueue.async {
            work()
            self.incrementCompletedCount()
        }
    }
    
    // MARK: - Async/Await Support
    
    /// Executes a background task asynchronously.
    @available(iOS 13.0, tvOS 13.0, *)
    public func executeBackgroundAsync<T>(_ work: @escaping () throws -> T) async throws -> T {
        incrementTaskCount()
        return try await withCheckedThrowingContinuation { continuation in
            backgroundQueue.addOperation {
                do {
                    let result = try work()
                    self.incrementCompletedCount()
                    continuation.resume(returning: result)
                } catch {
                    self.incrementCompletedCount()
                    continuation.resume(throwing: error)
                }
            }
        }
    }
    
    /// Executes a network task asynchronously.
    @available(iOS 13.0, tvOS 13.0, *)
    public func executeNetworkAsync<T>(_ work: @escaping () throws -> T) async throws -> T {
        incrementTaskCount()
        return try await withCheckedThrowingContinuation { continuation in
            networkQueue.addOperation {
                do {
                    let result = try work()
                    self.incrementCompletedCount()
                    continuation.resume(returning: result)
                } catch {
                    self.incrementCompletedCount()
                    continuation.resume(throwing: error)
                }
            }
        }
    }
    
    /// Executes a compute task asynchronously.
    @available(iOS 13.0, tvOS 13.0, *)
    public func executeComputeAsync<T>(_ work: @escaping () throws -> T) async throws -> T {
        incrementTaskCount()
        return try await withCheckedThrowingContinuation { continuation in
            computeQueue.addOperation {
                do {
                    let result = try work()
                    self.incrementCompletedCount()
                    continuation.resume(returning: result)
                } catch {
                    self.incrementCompletedCount()
                    continuation.resume(throwing: error)
                }
            }
        }
    }
    
    // MARK: - Statistics
    
    /// Gets current pool statistics.
    public func getStats() -> [String: Any] {
        metricsLock.lock()
        let tasks = taskCount
        let completed = completedTaskCount
        metricsLock.unlock()
        
        return [
            "cpuCores": cpuInfo.totalCores,
            "isHeterogeneous": cpuInfo.isHeterogeneous,
            "processorType": cpuInfo.processorType,
            "backgroundQueue": [
                "maxConcurrent": backgroundQueue.maxConcurrentOperationCount,
                "operationCount": backgroundQueue.operationCount
            ],
            "networkQueue": [
                "maxConcurrent": networkQueue.maxConcurrentOperationCount,
                "operationCount": networkQueue.operationCount
            ],
            "computeQueue": [
                "maxConcurrent": computeQueue.maxConcurrentOperationCount,
                "operationCount": computeQueue.operationCount
            ],
            "totalTasksSubmitted": tasks,
            "totalTasksCompleted": completed
        ]
    }
    
    // MARK: - Queue Management
    
    /// Cancels all pending operations.
    public func cancelAllOperations() {
        backgroundQueue.cancelAllOperations()
        networkQueue.cancelAllOperations()
        computeQueue.cancelAllOperations()
    }
    
    /// Waits until all operations are finished.
    public func waitUntilAllOperationsAreFinished() {
        backgroundQueue.waitUntilAllOperationsAreFinished()
        networkQueue.waitUntilAllOperationsAreFinished()
        computeQueue.waitUntilAllOperationsAreFinished()
    }
    
    /// Suspends all queues.
    public func suspendQueues() {
        backgroundQueue.isSuspended = true
        networkQueue.isSuspended = true
        computeQueue.isSuspended = true
    }
    
    /// Resumes all queues.
    public func resumeQueues() {
        backgroundQueue.isSuspended = false
        networkQueue.isSuspended = false
        computeQueue.isSuspended = false
    }
    
    // MARK: - Private Helpers
    
    private func incrementTaskCount() {
        metricsLock.lock()
        taskCount += 1
        metricsLock.unlock()
    }
    
    private func incrementCompletedCount() {
        metricsLock.lock()
        completedTaskCount += 1
        metricsLock.unlock()
    }
    
    private static func detectCpuInfo() -> CpuInfo {
        let totalCores = ProcessInfo.processInfo.processorCount
        let activeProcessors = ProcessInfo.processInfo.activeProcessorCount
        
        // Detect processor type
        var systemInfo = utsname()
        uname(&systemInfo)
        let machine = withUnsafePointer(to: &systemInfo.machine) {
            $0.withMemoryRebound(to: CChar.self, capacity: 1) {
                String(validatingUTF8: $0) ?? "Unknown"
            }
        }
        
        // Check for Apple Silicon (ARM-based chips have heterogeneous cores)
        let isAppleSilicon = machine.contains("arm") || machine.hasPrefix("iPhone") || 
                            machine.hasPrefix("iPad") || machine.hasPrefix("AppleTV")
        
        // Apple Silicon Macs and iOS devices typically have heterogeneous cores
        // We estimate the split based on typical configurations
        var performanceCores = totalCores
        var efficiencyCores = 0
        var isHeterogeneous = false
        
        if isAppleSilicon {
            isHeterogeneous = true
            // Typical Apple Silicon configurations:
            // A14+: 2 performance + 4 efficiency
            // M1: 4 performance + 4 efficiency
            // M1 Pro/Max: 8 performance + 2 efficiency
            // We'll estimate performance cores as ~50% of total
            performanceCores = max(2, totalCores / 2)
            efficiencyCores = totalCores - performanceCores
        }
        
        return CpuInfo(
            totalCores: totalCores,
            performanceCores: performanceCores,
            efficiencyCores: efficiencyCores,
            isHeterogeneous: isHeterogeneous,
            processorType: machine
        )
    }
}

// MARK: - OperationQueue Extension for Convenience

extension OperationQueue {
    /// Creates an operation that will run the given work.
    public func addAsyncOperation(_ work: @escaping () async throws -> Void) {
        let operation = BlockOperation()
        operation.addExecutionBlock {
            let semaphore = DispatchSemaphore(value: 0)
            Task {
                do {
                    try await work()
                } catch {
                    // Log error if needed
                }
                semaphore.signal()
            }
            semaphore.wait()
        }
        addOperation(operation)
    }
}
